ci: use zram instead of on-disk swap

thunder-coding opened the pull request.

thunder-coding committed:
ci(packages): use zram for large packages

Should improve performance when compiling large packages. And zram
should be unused unless there is requirement for it's usage. Not sure
how much will it affect packages that do not need additional memory
(needs to be monitored for CI usage after getting merged)

thunder-coding committed:
ci(packages): apply changes requested by robert

Changes made:
- download tarball from cdn.kernel.org instead of git clone
  Also check the signature of the tarball

- Manual install module instead of using modules_install
  modules_install does do the same thing and run depmod. We are running
  depmod ourselves. Also kernel Makefile will skip running depmod since
  System.map is not available on the VM

- Also do zstd compression on the module. The kernel makefiles don't
  seem to do it by default

thunder-coding committed:
ci(packages): introduce github actions caching for zram module

- Store some of the common variables as outputs for reuse

- Manually restore and save cached zram.ko.zst. This should ensure that
  the builds are cached even when building packages fails

- Additionally the built zram.ko.zst needs to be stored in scripts/ as
  GHA doesn't support caching absolute paths well:
  See issues:
  - https://github.com/actions/cache/issues/1127
  - https://github.com/actions/cache/issues/1540 which was worked around
    in
    https://github.com/oro-os/toolchain/commit/ebcad88ba6e7367643ba805071e77d14ca677e70
    by using relative path
  Also restoring cache in /lib/modules/ would require
  actions/cache/restore to run as root, and exists no documented way to
  do that which I could find online

thunder-coding commented:
Thanks! `git commit --squash` worked pretty well (except for the mess-ups I created myself). Was able to reorder some commits as well pretty well. Not sure why stacked PRs are so much in trend in the industry nowadays than just relying on this feature which is built into git itself

thunder-coding commented on .github/workflows/packages.yml:
```diff
@@ -45,6 +45,65 @@ jobs:
+        key: ${{ steps.kernel-info.outputs.kernel_version }}-${{ hashFiles('scripts/linux-kernel-signing-keys.gpg') }}-${{ hashFiles('.github/workflows/packages.yml') }}
+
+    - name: Build zram kernel module
+      if: steps.cache-zram.outputs.cache-hit != 'true'
```
The paths specified by `path` are cached by GHA. If some data is available with the specified key, then it is restored (happens above in `actions/cache/restore` step).

We only need to build the module if we couldn't restore from cache.

We further save the cache just after we build the module and load it.

thunder-coding committed:
ci(packages): strict mode for all zram scripts

thunder-coding commented:
This PR should be ready for review. Some action runs:
- https://github.com/termux/termux-packages/actions/runs/20459568208?pr=27733

Older runs before cleaning commits showing that caching of the built DKMS module works:
- https://github.com/termux/termux-packages/actions/runs/20458562801/job/58786018803?pr=27733 (DKMS module was built here)
- https://github.com/termux/termux-packages/actions/runs/20459325908/job/58788341889?pr=27733 (Prebuilt module used from cache)

robertkirkman commented on .gitignore:
```diff
@@ -43,3 +43,9 @@ checksum*
+# Generated on GitHub Actions CI when building kernel modules
+# This needs to be here so that GitHub Actions can cache the module
+# without having to run it as root (which I couldn't find any
+# documentation about)
```
Ok, thanks for explaining! In summary, it seems like the concise way to explain this could be "Since `actions/cache` does not support absolute paths, the `zram.ko.zst` needs to be cached in a relative path under `${{ github.workspace }}`, which should be excluded from Git using `.gitignore`".

robertkirkman commented:
I almost forgot, will this be in `package_updates.yml` as well as `packages.yml`?

thunder-coding committed:
ci: enable zram for package_updates using composite actions

The earlier logic for zram has been extracted into .github/actions/zram
for reuse as composite action.
https://docs.github.com/en/actions/tutorials/create-actions/create-a-composite-action

thunder-coding committed:
chore,ci(packages_updates): pin ubuntu version to the one in packages
workflow

Should reduce any unexpected surprised with things working in package
updates but failing when the commit is pushed to master. ubuntu-latest
is ubuntu-24.04 at the time of this commit. So this has no effect
whatsoever in any way

thunder-coding commented:
> I almost forgot, will this be in `package_updates.yml` as well as `packages.yml`?

Thanks, 66b823280e66cc82ac57e791e04f5e9ec7fbbc06 should address this.
